### ISSUE TITLE
very poor performance of graalvm when compared to JNI

### ISSUE THREAD

**Issue by thawaria** [oracle/graal#605]
Hi,
I am trying to call same C++ project using GraalVm and JNI. I am getting very poor performance using GraalVM (100 times lower).

Below is my code for GraalVm
Encode.java

```
import java.io.*;
import java.util.function.Function;
import java.nio.charset.Charset;
import org.graalvm.*;

public class Encoder {
	public static void main(String[] args) {
		String input = "This is a list of Unicode characters. As of version 11.0, Unicode contains a repertoire of over 137,000 characters covering 146 modern and historic scripts, as well as multiple symbol sets. As it is not technically possible to list all of these characters in a single Wikipedia page, this list is limited to a subset of the most important characters for English-language readers, with links to other pages which list the supplementary characters. This page includes the 1062 characters in the Multilingual European Character Set 2 (MES-2) subset, and some additional related characters.";
 		try (Context context = Context.newBuilder().allowNativeAccess(true).build()) {
			final Value library;
			try {
				library = context.eval(Source.newBuilder("llvm", new File("aztec.bc")).build());
				for(int i = 0;i<100;i++) {
					long start = System.nanoTime();
					final Value main = library.getMember("EncodeOptimised");
					String result = main.execute(input, input.length(), Charset.defaultCharset().toString(), (Function<Integer, char[]>) s -> new char[s]).toString();
					long timeElapsed = System.nanoTime() - start;
					System.out.println(result);
					System.out.println("\n=== time elapsed in milliseconds = "+ timeElapsed/1000000);
				}
			}
			catch (Exception e) {
			}}}}
```

Code for JNI
EncodeJNI.java

  ```
 public class EncoderJNI {
    static {
      System.loadLibrary("aztec"); // Load native library at runtime
    }
    private native String encodeNative(String msg);
    
     public static void main(String[] args) {
        EncoderJNI ref = new EncoderJNI();
	String input = "This is a list of Unicode characters. As of version 11.0, Unicode contains a repertoire of over 137,000 characters covering 146 modern and historic scripts, as well as multiple symbol sets. As it is not technically possible to list all of these characters in a single Wikipedia page, this list is limited to a subset of the most important characters for English-language readers, with links to other pages which list the supplementary characters. This page includes the 1062 characters in the Multilingual European Character Set 2 (MES-2) subset, and some additional related characters.";
   
        for(int i = 0;i<100;i++) 
	{
		long start = System.nanoTime();
         	String result  = ref.encodeNative(input);
		long timeElapsed = System.nanoTime() - start;
		System.out.println(result);
		System.out.println("\n=== time elapsed in milliseconds = "+ timeElapsed/1000000);
	}    }  }
```
I am using `graalvm-ee-1.0.0-rc4` and `-O3` level to compile native code.
Is this expected behavior? Can I do something to improve the performance?

**Comment by thawaria** [comment]
Thanks for the reply. I have found the issue. 
Debug symbol were not stripped(-g option was not removed) while generating LLVM IR.
After stripping debug information Graalvm is giving comparable performance numbers.
but It should not give 100 times slower performance even after allowing debug symbol.

I have one more query regarding how to run GraalVM in memory managed mode? 
Does GraalVM manage memory leak as well in the native code?

**Comment by rschatz** [comment]
> Debug symbol were not stripped(-g option was not removed) while generating LLVM IR.
After stripping debug information Graalvm is giving comparable performance numbers.
but It should not give 100 times slower performance even after allowing debug symbol.

In theory, debug info should not make any performance difference at all. But we have one known bug that is consistent with what you're seeing. In some rare circumstances, debug info can introduce a corner case where we infinitely deopt and compile again without making real progress. We're already working on a fix for that.

> I have one more query regarding how to run GraalVM in memory managed mode?
Does GraalVM manage memory leak as well in the native code?

No, the LLVM backend uses regular native allocations that are not managed by the garbage collector.

**Comment by preetnitkkr** [comment]
Hi Ronald,

> No, the LLVM backend uses regular native allocations that are not managed by the garbage collector.
I guess by LLVM backend, you are referring to Sulong. Please correct me if that's not the case. 

Also, the paper https://dl.acm.org/citation.cfm?doid=2786558.2786565 - talks about the TruffleC and ManagedC implementations. By your comment, it seems that the LLVM backend is using TruffleC implementation. Is there any way to switch to ManagedC implementation under relaxed mode ?

Thanks !

**Comment by rschatz** [comment]
This paper is about an interpreter of the C language directly, as opposed to Sulong, which is an interpreter for LLVM bitcode. We are exploring ways to bring memory safety to Sulong, but none of that is publicly available at the moment.

**Comment by preetnitkkr** [comment]
> We are exploring ways to bring memory safety to Sulong,

In paper. http://ssw.jku.at/General/Staff/ManuelRigger/ECOOP16-DS.pdf  , titled "Sulong: **Memory Safe** and Efficient Execution of LLVM-Based Languages", it is written in various sections that Sulong provides complete memory safety (and 2 other benefits to make it adaptable to production level). Also, there have been in-depth discussions on various techniques to achieve the safety.

By your comment, it seems that the memory safety work is currently work in progress and is not public at the moment.

Basically, we at our company are evaluating the GraalVM for a C/C++ legacy code we have to take to server side on production level. But since memory is currently unmanaged and hence can lead to JVM crash in case of memory errors and ultimately server down, it seems like running of native code on GraalVM will still take some time to reach the production level :( 

Could you please give a very rough estimate of how long can it take ? That will help us further in evaluating future paths of our project.

Thanks !

**Comment by rschatz** [comment]
The current plan is to push out a first preview of our work on memory safety end of September.

**Comment by preetnitkkr** [comment]
Thank you so much, Ronald. It helped a lot !

**Comment by thawaria** [comment]
Thanks for the information, Ronald.

**Comment by rschatz** [comment]
We have just published an experimental option for making LLVM on GraalVM memory-safe. See https://medium.com/graalvm/safe-and-sandboxed-execution-of-native-code-f6096b35c360 for more details.
